Provide Specs for the Standard Library

juliand665 opened the pull request:
As part of my thesis under @Aurel300, this pull request adds specifications to the most popular types & functions in the Rust Standard Library. I am using data gathered through [qrates](https://github.com/rust-corpus/qrates) to inform decisions on what specs to prioritize, along with a sense of logical units, e.g. covering all members of the unconstrained `Option<T>` or `Result<T, E>` `impl`s, rather than only specifying the subset of these methods that sees the most usage.

## Outline of Planned & Completed Specifications

These specs are very much a work in progress and far from complete at the moment, though what is there should be sound.

### `core::option::Option`
- [x] main impl

### `core::result::Result`
- [x] main impl

### `core::clone::Clone`
- [x] snapshot equality with opt-out mechanism

### `core::default::Default`
- [x] purity (for usage in specs) with opt-out mechanism

#### specific default values
- [x] `()`, `bool`, `char`, numbers (the types specified [here](https://github.com/rust-lang/rust/blob/872631d0f0fadffe3220ab1bd9c8f1f2342341e2/library/core/src/default.rs#L204-L223))
- [x] other commonly-used types

### `core::ops::Deref`
- [ ] purity (for usage in specs) with opt-out mechanism

This is currently blocked due to #1221.

### `core::ops::Index`/`core::ops::IndexMut`
- [ ] purity (for usage in specs) with opt-out mechanism

This is similarly blocked due to #1221.

### `core::ops::Try`
- [x] impl for `core::option::Option`
- [x] impl for `core::option::Result`

### `core::mem::size_of`
- [x] trait to allow users to specify what this method will return for their type
- [x] via the above, numbers, which are [by far](https://gist.github.com/juliand665/8948233f08f795b7eaade82076ea4fde) the most commonly-passed types

### `core::convert::From`/`Into`
- [x] purity (for usage in specs) with opt-out mechanism
- [x] numeric conversions, which are again [very common](https://gist.github.com/juliand665/486ec64a985de37675dcfd152f2b2bce)

### Slices/Arrays
- [x] indexing with `usize` (built-in)
- [x] length (built-in)
- [ ] "unsizing" arrays to slices

Indexing is currently blocked due to #1221.

### `alloc::vec::Vec`
- [ ] `Deref` as slice (#1221)
- [ ] forwarding of most operations to `as_slice` (#1221)
- [ ] `vec!` macro (requires unsizing from above, but very simple given that!)

### `alloc::string::String`
- [ ] `Deref` as `str` (#1221)
- [ ] forwarding of most operations to `as_str` (#1221)
- [ ] `&str` to `String` conversion using `From`/`Into`
I envision Strings and Vecs to work very similarly, mostly powered by a pure `as_str`/`as_slice` that acts as the source of truth. Unfortunately, that approach (like so much other stuff) is currently blocked, but I think the issue is so fundamental that it's better to just resolve it directly than to find a temporary workaround.

### `core::cmp::PartialEq`/`core::cmp::PartialOrd`
These are currently blocked due to #1311.

### Ranges
- [ ] `contains` on inclusive & half-open ranges, for loops & invariants (#1311)

### `core::ops` Binary Ops for References
- [x] specify how these ops are equivalent to applying to deref'd values
- [ ] the same for comparisons (#1311)

These operations don't have blanket impls for when one or both sides are a reference, but their implementations are [unified using macros](https://github.com/rust-lang/rust/blob/fd815a5091eb4d49cd317f8ad272f17b7a5f550d/library/core/src/ops/arith.rs#L110), which probably makes sense for us to replicate.

### Smart Pointers (e.g. `alloc::rc::Rc`)
This fundamentally relies on `Deref` and is thus also blocked by #1221.
Once that is resolved, specifying their `Deref` implementation as pure and expressing transfer into and out of a box via e.g. `ensures(old(x) === snap(result.deref()))` should go a long way towards specifying this type usefully.
Note that `Box` already has builtin support in Prusti, partly because Rust itself already treats it specially (e.g. that `*box` can move out of the box).

juliand665 committed:
initial specs

juliand665 committed:
add specs for try trait

juliand665 committed:
Merge remote-tracking branch 'upstream/master' into prusti-std

juliand665 committed:
specify some known sizes & alignments

juliand665 committed:
specify swap

Aurel300 commented on prusti-contracts/prusti-contracts/src/core_spec/default.rs:
```diff
@@ -0,0 +1,47 @@
+            #[ensures(result == $v)]
+            fn default() -> Self {
+                $v
+            }
```
This should not have a body.

juliand665 replied:
As soon as I switch away from my branch with the change to warn on bodies for such methods lol, good catch.

Aurel300 commented on prusti-contracts/prusti-contracts/src/core_spec/mem.rs:
```diff
@@ -0,0 +1,65 @@
+        fn align_of<T>() -> usize;
+		
+		#[ensures(*x === old(snap(y)) && *y === old(snap(x)))]
+		fn swap<T>(x: &mut T, y: &mut T);
```
Use spaces for indentation throughout.

Aurel300 commented on prusti-contracts/prusti-contracts/src/core_spec/mem.rs:
```diff
@@ -0,0 +1,65 @@
+        #[refine_trait_spec]
+        impl KnownSize for $t {
+            #[pure]
+            #[ensures(result == $size)]
```
This postcondition should be unnecessary, same for `align` below. (The body of a pure function becomes part of its contract…)

juliand665 replied:
Ah, I was so used to using it in extern specs that that wasn't in my mental model.

juliand665 committed:
formatting

juliand665 committed:
remove invalid extern spec body

Aurel300 commented on prusti-contracts/prusti-contracts/src/core_spec/mem.rs:
```diff
@@ -0,0 +1,65 @@
+known_size_spec!(f32, 4, 4);
+known_size_spec!(f64, 8, 8);
+
+// usize/isize are not specified exactly because they are platform-dependent and programmers should not depend on their specific values anyway.
```
We could say that the value is one of the three possible for Rust, i.e. [16-, 32- or 64-bit](https://doc.rust-lang.org/src/core/num/mod.rs.html#243-268). However, I think we can do even better: put the `known_size_spec!` under a `cfg`, so when compiling the correct value is chosen.

juliand665 replied:
Yeah, I considered something like that. This would allow you to ensure that your code is correct across different word sizes without having to run Prusti for each platform, while specifying the exact size gives information the programmer should maybe not be relying on? But I can absolutely specify it if you think it makes sense.

Aurel300 replied:
I think we make the same assumption (that the verification machine = the target machine) when processing MIR [`SizeOf`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/enum.NullOp.html#variant.SizeOf) operations. So for now, I would suggest using the `cfg` approach to be consistent. In the future we might think about "cross-verification".

Aurel300 commented on prusti-contracts/prusti-contracts/src/core_spec/clone.rs:
```diff
@@ -0,0 +1,12 @@
+trait Clone {
+    #[ghost_constraint(Self: SnapshotEqualClone, [
+		ensures(result === self),
+	])]
```
Use spaces throughout.

juliand665 replied:
Ah right, rustfmt doesn't touch whitespace within attributes because it treats them opaquely. I'll find-replace all tabs with 4 spaces, and try to keep in mind to check occasionally since they seem to like to slip in…

juliand665 committed:
remove unnecessary specs on pure functions

Aurel300 commented on prusti-contracts/prusti-contracts/src/core_spec/result.rs:
```diff
@@ -0,0 +1,181 @@
+
+#[extern_spec]
+impl<T, E> Result<T, E> {
+    // TODO: already in prusti-contracts
```
Stale TODO?

juliand665 committed:
remove resolved TODO

juliand665 committed:
attribute formatting

rustfmt can't handle this

juliand665 committed:
add shorthand for when size = alignment

juliand665 committed:
specify size of usize/isize

juliand665 committed:
ever more formatting

juliand665 committed:
Merge branch 'master' into prusti-std

juliand665 committed:
update specs to use merged improvements

juliand665 committed:
include try trait specs

juliand665 committed:
specify binop reference forwarding

juliand665 committed:
add compiletests for built-in specs

more coming soon

Aurel300 commented on prusti-contracts/prusti-contracts/src/core_spec/default.rs:
```diff
@@ -0,0 +1,45 @@
+/// Specifies that `Default::default`, if implemented, is a pure method, allowing its usage in specs.
+///
+/// Does not apply to types that do not implement `Copy`, since pure methods can only involve `Copy` types.
+pub auto trait PureDefault {}
```
I don't know if I asked before: can `PureDefault` be declared with `Copy` as a super-trait? Or does that not work for auto traits?

juliand665 replied:
They unfortunately cannot have super traits, I just checked again to make sure. The explanation sounds like it's not planned either, rather a principle I'd disagree with.

<img width="468" alt="image" src="https://user-images.githubusercontent.com/827762/212780466-39f8b495-736f-4872-92f4-efd0984d311b.png">

Aurel300 commented on prusti-contracts/prusti-contracts/src/core_spec/default.rs:
```diff
@@ -0,0 +1,45 @@
+    };
+}
+
+default_spec! { (), () }
```
Can we also specify the `Default` for non-unit tuples? [It exists for up to 12-item tuples.](https://doc.rust-lang.org/stable/std/primitive.tuple.html#impl-Default-for-(T%2C))

Aurel300 commented on prusti-tests/tests/compiletest.rs:
```diff
@@ -198,6 +198,11 @@ fn test_runner(_tests: &[&()]) {
     save_verification_cache();
 
+    // Test the functionality of built-in specs.
+    println!("[prusti_std]");
```
(As messaged) I would name this directory simply `std`.

juliand665 committed:
include generic params in ghost constraint evaluation substs

need this for tuple default specs, and probably others

juliand665 committed:
update tests to use included specs

juliand665 committed:
more spec removal

juliand665 committed:
add specs for convert::From and convert::Into

juliand665 committed:
specify Default behavior for tuples

juliand665 committed:
rename std tests folder

juliand665 committed:
minor tweaks

juliand665 committed:
basic slice/vec/str/string spec framework

very basic due to blockage by #1221

juliand665 committed:
allow declaring alloc as extern crate

this is an exception listed here: https://doc.rust-lang.org/edition-guide/rust-2018/path-changes.html

juliand665 committed:
move alloc specs to prusti-std crate

juliand665 committed:
minor fixes

juliand665 committed:
update test to use core spec

juliand665 committed:
add future spec

juliand665 committed:
Merge remote-tracking branch 'upstream/master' into prusti-std

juliand665 committed:
minor tweaks

juliand665 committed:
option flatten & option/result transpose

Pointerbender commented on prusti-contracts/prusti-std/src/collections.rs:
```diff
@@ -0,0 +1,21 @@
+    K: Eq + ::core::hash::Hash,
+    S: ::std::hash::BuildHasher,
+{
+    #[pure]
```
Is marking this generic method as `#[pure]` also sound in the presence of interior mutability? (e.g. what if a hypothetical `PartialEq::eq` implementation modifies its internals?)

fpoli replied:
That's a good point, I don't think it would be. It's also related to https://github.com/viperproject/prusti-dev/issues/1286#issuecomment-1406247838.

I wonder if with `refine_spec` we can restrict all these specifications to types `K/Q` for which `Eq/Hash/Borrow` is pure.

Pointerbender replied:
Would that also work if `K` is replaced by `&K` which then delegates its impl to a potentially non-structural impl?

fpoli replied:
(It would be nice to add to this PR a test where a key type implements a random `PartialEq` and `Hash`. A client of the `HashMap` should then *not* be able to assert anything about the result of `HashMap::get`.)

fpoli replied:
> Would that also work if `K` is replaced by `&K` which then delegates its impl to a potentially non-structural impl?

I edited my message. The issue is not just interior mutability, but also non-determinism.

Aurel300 replied:
> I wonder if with `refine_spec` we can restrict all these specifications to types `K/Q` for which `Eq/Hash/Borrow` is pure.

We could.

I think `get` being deterministic is not controversial for a `HashMap`.

However, I think/hope interior mutability isn't actually such a big problem. In my mental model, an interior-mutable structure can be expected to be snapshot equal. As we encode its snapshot recursively, we will eventually find the `UnsafeCell`, which is where we stop: the cell is a *handle*, which does not tell us anything about the value until we also have e.g. a heap (version) to refer to.

juliand665 committed:
specify size of arrays

juliand665 committed:
conditionalize std specs

juliand665 committed:
add missing import

juliand665 committed:
add tests for clone specs

includes inheritance of postconditions and auto trait "transitivity"

juliand665 committed:
use snapshot equality for Default specs

have to desugar this manually since prusti syntax breaks rust's macro parsing

juliand665 committed:
update smir checks to allow what prusti-std needs

juliand665 committed:
add missing "supertrait" bounds

juliand665 committed:
comment out all #1221 specs for now

juliand665 committed:
fix warning

Aurel300 commented on prusti-tests/tests/verify_overflow/pass/extern-spec/vec-3.rs:
```diff
@@ -1,28 +1,12 @@
 impl<T> Vec<T> {
     #[ensures(result.len() == 0)]
-    fn new() -> Vec::<T>;
+    fn new() -> Vec<T>;
```
`prusti-contracts/prusti-std/src/alloc_spec/vec.rs` also defines a spec for `new`. Why is this not considered a duplicate?

Aurel300 replied:
Because in these tests only the `core` specs are used, `std` specs require the `prusti-std` crate as a dependency.

Aurel300 committed:
Merge branch 'master' into prusti-std

vakaras committed:
Comment out failing test.

Aurel300 committed:
bump versions

Aurel300 committed:
Merge remote-tracking branch 'upstream/master' into prusti-std

Aurel300 committed:
Merge remote-tracking branch 'upstream/master' into prusti-std

Aurel300 committed:
account for prusti-std in user guide

Aurel300 committed:
Merge remote-tracking branch 'upstream/master' into prusti-std
